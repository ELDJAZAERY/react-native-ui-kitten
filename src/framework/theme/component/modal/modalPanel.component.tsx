import React from 'react';
import {
  View,
  StyleSheet,
  ViewProps,
} from 'react-native';
import {
  ModalAnimationConfig,
  ModalService,
} from '../../service';
import { Modal } from '../../../ui/modal/modal.component';

export interface ModalPanelProps {
  children: React.ReactElement<any> | React.ReactElement<any>[];
}

interface ModalPanelState {
  dialogComponents: Map<string, React.ReactElement<any>>;
  closeOnBackDropAllowed: boolean;
  modalAnimationConfig: ModalAnimationConfig;
}

export class ModalPanel extends React.Component<ModalPanelProps, ModalPanelState> {

  public state: ModalPanelState = {
    dialogComponents: new Map(),
    closeOnBackDropAllowed: false,
    modalAnimationConfig: {
      animationType: 'none',
      animationDuration: 0,
    },
  };

  public componentDidMount(): void {
    ModalService.setComponent(this);
  }

  public componentWillUnmount(): void {
    ModalService.setComponent(null);
  }

  public onCloseModal = (identifier: string) => {
    const components: Map<string, React.ReactElement<any>> = this.state.dialogComponents;
    components.delete(identifier);
    this.setState({
      dialogComponents: components,
      modalAnimationConfig: {
        animationType: 'none',
        animationDuration: 0,
      },
    });
  };

  public showDialog(dialogComponent: React.ReactElement<any>,
                    closeOnBackDrop: boolean,
                    animationConfig?: ModalAnimationConfig): void {
    const map: Map<string, React.ReactElement<any>> = this.state.dialogComponents
      .set(this.generateUniqComponentKey(), dialogComponent);
    const animConfig: ModalAnimationConfig = animationConfig ?
      animationConfig : this.state.modalAnimationConfig;
    this.setState({
      dialogComponents: map,
      closeOnBackDropAllowed: closeOnBackDrop,
      modalAnimationConfig: animConfig,
    });
  }

  private generateUniqComponentKey = (): string => {
    return Math.random().toString(36).substring(2);
  };

  private areThereAnyComponents(): boolean {
    return this.state.dialogComponents && this.state.dialogComponents.size !== 0;
  }

  private renderModal(modal: React.ReactElement<any>, index: number) {
    const allModalKeys: string[] = Array.from(this.state.dialogComponents.keys());
    const identifier: string = allModalKeys
      .find(item => this.state.dialogComponents.get(item) === modal);
    return (
      <Modal
        visible={true}
        isBackDropAllowed={this.state.closeOnBackDropAllowed}
        key={index}
        identifier={identifier}
        animationType={this.state.modalAnimationConfig.animationType}
        animationDuration={this.state.modalAnimationConfig.animationDuration}
        onCloseModal={this.onCloseModal}
      >
        {modal}
      </Modal>
    );
  }

  private renderModals() {
    return Array.from(this.state.dialogComponents.values())
      .map((component: React.ReactElement<any>, i: number) =>
        this.renderModal(component, i));
  }

  public render(): React.ReactElement<ViewProps> {
    return (
      <View style={styles.container}>
        {this.props.children}
        {this.areThereAnyComponents() && this.renderModals()}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
  },
});
