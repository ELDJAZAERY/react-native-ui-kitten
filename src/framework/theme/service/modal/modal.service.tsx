import React from 'react';
import { ModalPanel } from '../../component';

export type ModalAnimationType = 'slideInUp' | 'fade' | 'none';

export interface ModalAnimationConfig {
  animationType: ModalAnimationType;
  animationDuration: number;
}

class ModalServiceType {

  component: ModalPanel | null = null;

  public setComponent(component: ModalPanel | null): void {
    this.component = component;
  }

  public showDialog(dialogComponent: React.ReactElement<any>,
                    closeOnBackDrop: boolean = false,
                    animationConfig?: ModalAnimationConfig): void {
    if (this.component) {
      this.component.showDialog(dialogComponent, closeOnBackDrop, animationConfig);
    }
  }
}

export const ModalService = new ModalServiceType();
