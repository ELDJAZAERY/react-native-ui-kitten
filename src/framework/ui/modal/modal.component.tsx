import React from 'react';
import {
  View,
  ViewProps,
  TouchableWithoutFeedbackProps,
  StyleSheet,
  TouchableWithoutFeedback,
  Dimensions,
} from 'react-native';
import {
  StyledComponentProps,
  StyleType,
} from '@kitten/theme';

interface ModalProps {
  visible: boolean;
  children: React.ReactElement<any> | React.ReactElement<any>[];
  isBackDropAllowed?: boolean;
  identifier?: string;
  onCloseModal?: (index: string) => void;
}

export type Props = StyledComponentProps & ViewProps & ModalProps;

export class Modal extends React.Component<Props> {

  static defaultProps: Partial<Props> = {
    visible: false,
    isBackDropAllowed: false,
  };

  private getComponentStyle = (style: StyleType): StyleType | null => ({
    container: style ? {
      paddingHorizontal: style.paddingHorizontal,
      paddingVertical: style.paddingVertical,
      backgroundColor: style.backgroundColor,
      borderColor: style.borderColor,
      borderRadius: style.borderRadius,
      borderWidth: style.borderWidth,
    } : null,
  });

  private closeModal: () => void = (): void => {
    this.props.onCloseModal && this.props.onCloseModal(this.props.identifier);
  };

  private closeOnBackDrop: () => void = () => {
    if (this.props.isBackDropAllowed) {
      this.closeModal();
    }
  };

  private renderBackdrop(): React.ReactElement<TouchableWithoutFeedbackProps> {
    return this.props.isBackDropAllowed ? (
      <TouchableWithoutFeedback onPress={this.closeOnBackDrop}>
        <View style={styles.backdrop}/>
      </TouchableWithoutFeedback>
    ) : null;
  }

  private renderComponent(): React.ReactElement<ViewProps> {
    const { children } = this.props;
    const componentStyle: StyleType = this.getComponentStyle(this.props.themedStyle);
    const childrenWithCloseProps = React.Children
      .map(children, (child: React.ReactElement<any>) =>
        React.cloneElement(child, {
            onCloseModal: this.closeModal,
            pointerEvents: 'box-none',
          },
        ));

    return (
      <View style={styles.container}>
        {this.renderBackdrop()}
        <View {...this.props} style={[componentStyle.container, this.props.style]}>
          {childrenWithCloseProps}
        </View>
      </View>
    );
  }

  public render(): React.ReactElement<ViewProps | TouchableWithoutFeedbackProps> {
    return this.props.visible ? this.renderComponent() : null;
  }
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    width: 0,
    height: 0,
  },
  backdrop: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height,
    ...StyleSheet.absoluteFillObject,
  },
});
