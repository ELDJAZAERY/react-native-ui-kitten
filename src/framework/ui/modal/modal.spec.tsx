import React from 'react';
import {
  fireEvent,
  render,
} from 'react-native-testing-library';
import {
  View,
  Text,
  TouchableWithoutFeedback,
  Button,
} from 'react-native';
import { Modal } from './modal.component';

describe('@modal component checks', () => {

  const MODAL_TEST_IDENTIFIER: (substring: string) => string = (substring: string): string =>
    `modal-test-identifier-${substring}`;

  it('* modal component renders properly', () => {
    const modal1 = render(<Modal
      visible={true}
      component={<View><Text>Test1</Text></View>}
      isBackDropAllowed={false}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={() => 1}/>);
    const modal2 = render(<Modal
      visible={false}
      component={<View><Text>Test2</Text></View>}
      isBackDropAllowed={false}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={() => 1}/>);

    expect(modal1).toMatchSnapshot();
    expect(modal2).toMatchSnapshot();
  });

  it('* modal component props checks', () => {
    const modalPassingProps = {
      visible: true,
      component: <View><Text>Test1</Text></View>,
      isBackDropAllowed: false,
    };
    const modal = <Modal {...modalPassingProps}/>;

    expect(modal.props.visible).toBe(modalPassingProps.visible);
    expect(modal.props.component).toBe(modalPassingProps.component);
    expect(modal.props.isBackDropAllowed).toBe(modalPassingProps.isBackDropAllowed);
  });

  it('* modal closes on passed prop', () => {
    const onCloseModal = jest.fn();
    const component = <Modal
      visible={true}
      component={<View>
        <Text>Test1</Text>
        <Button title={'Close Modal'} onPress={onCloseModal}/>
      </View>}
      isBackDropAllowed={true}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={onCloseModal}/>;
    const modal = render(component);
    expect(modal).toMatchSnapshot();
    fireEvent.press(modal.getByType(Button));
    expect(onCloseModal).toHaveBeenCalled();
    expect(modal).toMatchSnapshot();
  });

  it('* modal component close on backDrop checks', () => {
    const onCloseModal = jest.fn();
    const modal = render(<Modal
      visible={true}
      component={<View><Text>Test1</Text></View>}
      isBackDropAllowed={true}
      identifier={MODAL_TEST_IDENTIFIER('1')}
      onCloseModal={onCloseModal}/>);
    expect(modal).toMatchSnapshot();
    fireEvent.press(modal.getByType(TouchableWithoutFeedback));
    expect(modal).toMatchSnapshot();
    expect(onCloseModal).toHaveBeenCalled();
  });

});
