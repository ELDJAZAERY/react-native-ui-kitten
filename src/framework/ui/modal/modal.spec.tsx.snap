// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@modal component checks * modal closes on passed prop 1`] = `
<View
  accessible={true}
  component={
    <View>
      <Text>
        Test1
      </Text>
      <Button
        onPress={[MockFunction]}
        title="Close Modal"
      />
    </View>
  }
  identifier="modal-test-identifier-1"
  isBackDropAllowed={true}
  onCloseModal={[MockFunction]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      undefined,
    ]
  }
  visible={true}
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Close Modal
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`@modal component checks * modal closes on passed prop 2`] = `
<View
  accessible={true}
  component={
    <View>
      <Text>
        Test1
      </Text>
      <Button
        onPress={
          [MockFunction] {
            "calls": Array [
              Array [
                undefined,
              ],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        title="Close Modal"
      />
    </View>
  }
  identifier="modal-test-identifier-1"
  isBackDropAllowed={true}
  onCloseModal={
    [MockFunction] {
      "calls": Array [
        Array [
          undefined,
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      undefined,
    ]
  }
  visible={true}
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
    <View
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Close Modal
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`@modal component checks * modal component close on backDrop checks 1`] = `
<View
  accessible={true}
  component={
    <View>
      <Text>
        Test1
      </Text>
    </View>
  }
  identifier="modal-test-identifier-1"
  isBackDropAllowed={true}
  onCloseModal={[MockFunction]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      undefined,
    ]
  }
  visible={true}
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component checks * modal component close on backDrop checks 2`] = `
<View
  accessible={true}
  component={
    <View>
      <Text>
        Test1
      </Text>
    </View>
  }
  identifier="modal-test-identifier-1"
  isBackDropAllowed={true}
  onCloseModal={
    [MockFunction] {
      "calls": Array [
        Array [
          "modal-test-identifier-1",
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    }
  }
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      undefined,
    ]
  }
  visible={true}
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component checks * modal component renders properly 1`] = `
<View
  component={
    <View>
      <Text>
        Test1
      </Text>
    </View>
  }
  identifier="modal-test-identifier-1"
  isBackDropAllowed={false}
  onCloseModal={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "position": "absolute",
      },
      undefined,
    ]
  }
  visible={true}
>
  <View
    onCloseModal={[Function]}
  >
    <Text>
      Test1
    </Text>
  </View>
</View>
`;

exports[`@modal component checks * modal component renders properly 2`] = `null`;
